File: dbos_server_service/auth_service/src/schemas/departments.py
"""Схемы для управления отделами и их доступом к сервисам."""

from datetime import datetime

from pydantic import BaseModel, Field


class DepartmentCreate(BaseModel):
    """Тело `POST /departments`."""
    name: str = Field(description="Машинно-читаемое имя отдела (уникально).")
    display_name: str = Field(description="Человеческое название.")


class DepartmentUpdateRequest(BaseModel):
    """Тело `PATCH /departments/{id}`. Оба поля опциональны; пустое тело → 422.

    `name` (slug) — иммутабельный identity для аудита и логов, не правится
    через update; для смены имени отдел пересоздаётся.
    """
    display_name: str | None = Field(
        default=None,
        min_length=1,
        max_length=256,
        description="Человеческое название.",
    )
    description: str | None = Field(
        default=None,
        max_length=1024,
        description="Свободно-форматный текст-пояснение для UI-карточки.",
    )


class DepartmentResponse(BaseModel):
    """Отдел в ответе list/get эндпоинтов."""
    department_id: str
    name: str
    display_name: str
    description: str | None = None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class GrantServiceAccessRequest(BaseModel):
    """Тело `POST /departments/{department_id}/services`."""
    service_name: str = Field(description="Сервис, к которому даём отделу access.")


class ServiceAccessResponse(BaseModel):
    """Ответ grant-эндпоинта — текущее состояние связи (dept, service)."""
    department_id: str
    service_name: str
    enabled: bool
    granted_at: datetime | None = None


class HardDeleteDepartmentRequest(BaseModel):
    """Тело `DELETE /departments/{department_id}`. `reason` обязателен для compliance-аудита."""
    reason: str = Field(
        min_length=1,
        max_length=256,
        description="Человекочитаемое обоснование hard-delete'а (compliance).",
    )

File: emm/web_ui/src/api/auth/departments.ts
/**
 * Thin wrappers for `/departments/*` endpoints in auth_service.
 */

import { apiDelete, apiGet, apiPatch, apiPost } from "@/api/client";
import type {
  Department,
  DepartmentCreateRequest,
  DepartmentUpdateRequest,
} from "@/api/auth/types";

export interface ServiceAccessResponse {
  department_id: string;
  service_name: string;
  enabled: boolean;
  granted_at?: string | null;
}

// Backend `DepartmentResponse` отдаёт `department_id`; UI-тип Department хранит
// поле как `id`. Нормализуем на чтении, чтобы не править ~10 точек потребления.
interface BackendDepartment {
  department_id?: string;
  id?: string;
  name: string;
  description?: string | null;
  is_active?: boolean;
  user_count?: number;
  created_at: string;
  updated_at?: string;
}

function normalizeDepartment(d: BackendDepartment): Department {
  return {
    id: d.id ?? d.department_id ?? "",
    name: d.name,
    description: d.description ?? null,
    user_count: d.user_count ?? 0,
    created_at: d.created_at,
    updated_at: d.updated_at,
  };
}

export async function listDepartments(): Promise<Department[]> {
  const raw = await apiGet<BackendDepartment[]>("/auth/v1/departments");
  return raw.map(normalizeDepartment);
}

export async function createDepartment(
  body: DepartmentCreateRequest,
): Promise<Department> {
  const raw = await apiPost<BackendDepartment>("/auth/v1/departments", body);
  return normalizeDepartment(raw);
}

// PATCH принимает только `name` и/или `description`. Пустое тело backend
// отбивает 422 `EMPTY_UPDATE`.
export async function updateDepartment(
  departmentId: string,
  body: DepartmentUpdateRequest,
): Promise<Department> {
  const raw = await apiPatch<BackendDepartment>(
    `/auth/v1/departments/${encodeURIComponent(departmentId)}`,
    body,
  );
  return normalizeDepartment(raw);
}

export function deleteDepartment(
  departmentId: string,
  body: { reason: string },
): Promise<void> {
  return apiDelete<void>(`/auth/v1/departments/${departmentId}`, body);
}

// ── Department ↔ services access ────────────────────────────────────────────
//
// Backend (auth_service) даёт GET (список привязанных сервисов) + POST/DELETE.
// `listDepartmentServices` возвращает массив `service_name` активных грантов.
export function listDepartmentServices(
  departmentId: string,
): Promise<string[]> {
  return apiGet<string[]>(
    `/auth/v1/departments/${encodeURIComponent(departmentId)}/services`,
  );
}

export function grantServiceAccess(
  departmentId: string,
  serviceName: string,
): Promise<ServiceAccessResponse> {
  return apiPost<ServiceAccessResponse>(
    `/auth/v1/departments/${encodeURIComponent(departmentId)}/services`,
    { service_name: serviceName },
  );
}

export function revokeServiceAccess(
  departmentId: string,
  serviceName: string,
): Promise<void> {
  return apiDelete<void>(
    `/auth/v1/departments/${encodeURIComponent(departmentId)}/services/${encodeURIComponent(serviceName)}`,
  );
}
